### PR TITLE
[onert] Manage ruy_context on runtime

### PR DESCRIPTION
**Pull request by periannath**
- Create `ExternalContext` class which has ruy context
  - CPU `BackendConext` class has `ExternalContext`
  - KernelGenerator and FullyConnectedLayer has shared_ptr of `ExternalContext`
- Link ruy library to libbackend_cpu.so

ONE-DCO-1.0-Signed-off-by: JiHwan Yeo <jihwan.yeo@samsung.com>

---

For issue : #3318 
From draft #3436

### CODE REVIEW

File: runtime/onert/backend/cpu/BackendContext.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+                 std::shared_ptr<IKernelGenerator> kernel_gen = nullptr,
+                 std::shared_ptr<ITensorRegister> tensor_register = nullptr,
+                 std::shared_ptr<IOptimizer> optimizer = nullptr)
+      : onert::backend::BackendContext(backend, graph, tensor_builder, constant_initializer,
+                                       kernel_gen, tensor_register, optimizer),
+        _external_context(new ExternalContext)
+  {
+  }
+
+  std::shared_ptr<ExternalContext> external_context() { return _external_context; }
```
Comment (ragmani) [inline runtime/onert/backend/cpu/BackendContext.h:45]: Could you add a comment about https://github.com/Samsung/ONE/issues/3318#issuecomment-664100840?

Comment (periannath) [reply runtime/onert/backend/cpu/BackendContext.h:45]: @ragmani Thanks. I've added NOTE and TODO comments.

File: runtime/onert/backend/cpu/ExternalContext.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+public:
+  ExternalContext() : _ruy_context(new ruy::Context)
+  {
+    SetMaxNumThreads(onert::util::getConfigInt(onert::util::config::RUY_THREADS));
+#ifdef USE_RUY_GEMV
+    _ruy_context->cache_policy = ruy::kCacheLHSOnNarrowMul;
+#endif
+  }
+
+  void SetMaxNumThreads(int max_num_threads)
```
Comment (wateret) [inline runtime/onert/backend/cpu/ExternalContext.h:47]: Method names must start with small "s"

Comment (periannath) [reply runtime/onert/backend/cpu/ExternalContext.h]: @wateret Thanks. I've fixed it.

**Comment by wateret** [comment]
I may not fully understand the whole code, but I was not able to see the reason why `IExternalContext` should be part of `BackendContext` yet. I also looked at #3436 . Could you please explain why?

**Comment by periannath** [comment]
@wateret I just intended to manage `IExternalContext` somewhere inside ONE runtime. At first, I defined it as part of cpu `Backend` and it is moved onto cpu `BackendContext` after offline discussion with @ragmani https://github.com/Samsung/ONE/issues/3318#issuecomment-664081927. 

If you are talking about why runtime have to manage ruy context, defining ruy context as thread_local in cker creates many issues for multi-thread environment. One ruy context managed by one session will solve multi-thread issues.

**Comment by wateret** [comment]
@periannath Thank you for explanation, just one more thing to ask. Where does `BackendContext::external_context()` get called except backend context generation? And where does `IExternalContext::setMaxNumThreads` get called except `ExeternalContext` constructor?

**Comment by periannath** [comment]
@wateret `BackendContext::external_context()` is only called at backend context generation, and `IExternalContext::setMaxNumThreads` is only called at `ExternalContext` constructor.
`IExternalContext::setMaxNumThreads` can be used if user want to change number of threads used by ruy library between inference, but currently we don't have API to support that feature.

**Comment by wateret** [comment]
Would that mean if there is no need to call `setMaxNumThreads` from runtime core, `ExternalContext` may not necessarily be part of `BackendContext`?

**Comment by periannath** [comment]
@wateret As far as one runtime session has one `ExternalContext`, `ExternalContext` can be moved into another class. It is not related to `setMaxNumThreads`.
I used `BackendContext` because it seems proper class for now. Having `ExternalContext` inside runtime core would make `libonert_core.so` depends on ruy library. In current implementation, `libbackend_cpu.so` depends on ruy library.

**Pull request merged**